tests/difficulty_test.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+//
+
+package tests
+
+import (
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/consensus/ethash"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+	"math/big"
+	"strings"
+)
+
+type DifficultyTests map[string]difficultyTest
+
+//go:generate gencodec -type difficultyTest -field-override difficultyTestMarshaling -out gen_difficultytest.go
+
+type difficultyTest struct {
+	ParentTimestamp    *big.Int    `json:"parentTimestamp"`
+	ParentDifficulty   *big.Int    `json:"parentDifficulty"`
+	UncleHash          common.Hash `json:"parentUncles"`
+	CurrentTimestamp   *big.Int    `json:"currentTimestamp"`
+	CurrentBlockNumber uint64      `json:"currentBlockNumber"`
+	CurrentDifficulty  *big.Int    `json:"currentDifficulty"`
+}
+
+type difficultyTestMarshaling struct {
+	ParentTimestamp    *math.HexOrDecimal256
+	ParentDifficulty   *math.HexOrDecimal256
+	CurrentTimestamp   *math.HexOrDecimal256
+	CurrentDifficulty  *math.HexOrDecimal256
+	UncleHash          common.Hash
+	CurrentBlockNumber math.HexOrDecimal64
+}
```

Comment:
These types need to move to a separate non-test file so package tests can be built without `go test`.
`go test` is special because it builds files with the \_test.go suffix. These files are ignored when
building the package with `go install`.

Please structure it the same way as other test types:

* add `difficulty_test_util.go` and define DifficultyTest there
* DifficultyTest should have a Run method that takes a `params.ChainConfig` parameter, does the actual check and returns `error`
* Use DifficultyTest in `difficulty_test.go`